src/net/server/handlers/mul.rs

```diff
@@ -192,16 +197,243 @@ mod tests {
+            (OFFSET_HEADER_NAME.clone(), valid_offset.into()),
+            (HeaderName::from_static("datasize"), DATA_SIZE.into()),
+        ]);
+        let malformed_data_size_value_headers = HashMap::from([
+            (OFFSET_HEADER_NAME.clone(), valid_offset.into()),
+            (DATA_SIZE_HEADER_NAME.clone(), MALFORMED_DATA_SIZE.into()),
+        ]);
+
+        // malformed query_id
+        let req = build_malformed_req(
+            port,
+            malformed_query_id,
+            valid_step.clone(),
+            valid_target_helper,
+            valid_headers.clone(),
+            valid_body.clone(),
+        );
+        expect_res(req, StatusCode::UNPROCESSABLE_ENTITY).await;
+
+        // malformed identity
+        let req = build_malformed_req(
+            port,
+            valid_query_id,
+            valid_step.clone(),
+            malformed_target_helper,
+            valid_headers.clone(),
+            valid_body.clone(),
+        );
+        expect_res(req, StatusCode::UNPROCESSABLE_ENTITY).await;
+
+        // malformed offset header name
+        let req = build_malformed_req(
+            port,
+            valid_query_id,
+            valid_step.clone(),
+            valid_target_helper,
+            malformed_offset_name_headers,
+            valid_body.clone(),
+        );
+        expect_res(req, StatusCode::UNPROCESSABLE_ENTITY).await;
+
+        // malformed offset header value
+        let req = build_malformed_req(
+            port,
+            valid_query_id,
+            valid_step.clone(),
+            valid_target_helper,
+            malformed_offset_value_headers,
+            valid_body.clone(),
+        );
+        expect_res(req, StatusCode::BAD_REQUEST).await;
+
+        // malformed data-size header name
+        let req = build_malformed_req(
+            port,
+            valid_query_id,
+            valid_step.clone(),
+            valid_target_helper,
+            malformed_data_size_name_headers,
+            valid_body.clone(),
+        );
+        expect_res(req, StatusCode::UNPROCESSABLE_ENTITY).await;
+
+        // malformed data-size header value
+        let req = build_malformed_req(
+            port,
+            valid_query_id,
+            valid_step.clone(),
+            valid_target_helper,
+            malformed_data_size_value_headers,
+            valid_body.clone(),
+        );
+        expect_res(req, StatusCode::BAD_REQUEST).await;
+
+        // malformed body
+        let req = build_malformed_req(
+            port,
+            valid_query_id,
+            valid_step.clone(),
+            valid_target_helper,
+            valid_headers.clone(),
+            malformed_body,
+        );
+        expect_res(req, StatusCode::BAD_REQUEST).await;
+    }
+
+    fn poll<F, T>(f: &mut F) -> Poll<T>
+    where
+        F: Future<Output = T> + Unpin,
+    {
+        Pin::new(f).poll(&mut Context::from_waker(futures::task::noop_waker_ref()))
```

Comment:
you could use `poll_unpin` here because `F` is constrained to be `Unpin` and there is no need to pin it.

also this function could probably have a more appropriate name like `poll_once` maybe?

Reply:
i think `poll_once` specifically implies that it will consume the future, right? whereas `poll` acts in the same way that `Future::poll` acts.